### PR TITLE
Fixed #33766 -- Resolved FilteredRelation.condition at referencing time.

### PR DESCRIPTION
**Pull request by charettes**
A lot of the complexity that relates to `FilteredRelation.condition` resolving came from the fact that it was referencing the relation it was filtering against and not the filtered relation itself.

For example, given

```python
qs.annotate(
    adventure_books=FilteredRelation('books, condition=Q(books__category="adventure"))
)
```

_Translating_ the `condition` to be `Q(adventure_books__category="adventure")`  once the query has knowledge about `adventure_books` allows for the removal of the `reuse_with_filtered_relation` convoluted logic that fooled join resolving into resolving `books` to `adventure_books`.

### CODE REVIEW

File: django/db/models/sql/query.py [hunk 1]
```diff
@@ -73,6 +73,19 @@ def get_children_from_q(q):
             yield child
 
 
+def rename_prefix_from_q(prefix, replacement, q):
```
Comment (charettes) [inline django/db/models/sql/query.py:76]: This would need to be handle expressions as well but it should be relatively trivial to implement; basically all refs to `relation__` must be turned into `alias__` instead.

**Comment by charettes** [comment]
The full diff of the SQL change for the `filtered_relation` tests

```diff
--- test_exclude_relation_with_join (filtered_relation.tests.FilteredRelationTests):postgresql:01ae9d4ca9
+++ test_exclude_relation_with_join (filtered_relation.tests.FilteredRelationTests):postgresql:d7b1aba90a
@@ -4,5 +4,6 @@
                 "filtered_relation_author"."object_id"
 FROM "filtered_relation_author"
 INNER JOIN "filtered_relation_book" book_alice ON ("filtered_relation_author"."id" = book_alice."author_id"
-                                                   AND (NOT (UPPER(book_alice."title"::text) LIKE UPPER(%s))))
+                                                   AND (NOT (UPPER(book_alice."title"::text) LIKE UPPER(%s)
+                                                             AND book_alice."title"::text IS NOT NULL)))
 WHERE book_alice."id" IS NOT NULL
```

The `test_exclude_relation_with_join` seems more adequate to apply negation against handle null fields but it doesn't have an impact here (since `Book.title.null is False`) though so there might be something at play in terms of field resolving but we should add a test to make sure filtered conditions do perform adequate null handling.

```diff
--- test_multiple_times (filtered_relation.tests.FilteredRelationTests):postgresql:01ae9d4ca9
+++ test_multiple_times (filtered_relation.tests.FilteredRelationTests):postgresql:d7b1aba90a
@@ -5,7 +5,5 @@
 FROM "filtered_relation_author"
 INNER JOIN "filtered_relation_book" book_title_alice ON ("filtered_relation_author"."id" = book_title_alice."author_id"
                                                          AND (UPPER(book_title_alice."title"::text) LIKE UPPER(%s)))
-INNER JOIN "filtered_relation_book" T3 ON ("filtered_relation_author"."id" = T3."author_id"
-                                           AND (UPPER(T3."title"::text) LIKE UPPER(%s)))
 WHERE (book_title_alice."id" IS NOT NULL
-       AND T3."id" IS NOT NULL)
+       AND book_title_alice."id" IS NOT NULL)
```

This demonstrate that this patch changes the semantic of `FilteredRelation` so it's reuse between `filter` call. That was something hinted by Anssi as a likely desirable property (https://github.com/django/django/pull/7560#issuecomment-280312114) to opt-out of the `filter().filter()` behaviour when dealing with multi-valued relationships and have it be reused. Likely not something we want to change here though.

```diff
--- test_nested_m2m_filtered (filtered_relation.tests.FilteredRelationTests):postgresql:01ae9d4ca9
+++ test_nested_m2m_filtered (filtered_relation.tests.FilteredRelationTests):postgresql:d7b1aba90a
@@ -5,5 +5,7 @@
 LEFT OUTER JOIN "filtered_relation_author_favorite_books" ON ("filtered_relation_author"."id" = "filtered_relation_author_favorite_books"."author_id")
 LEFT OUTER JOIN "filtered_relation_book" T4 ON ("filtered_relation_author_favorite_books"."book_id" = T4."id"
                                                 AND (UPPER(T4."title"::text) LIKE UPPER(%s)))
+LEFT OUTER JOIN "filtered_relation_book" T5 ON ("filtered_relation_author_favorite_books"."book_id" = T5."id"
+                                                AND (UPPER(T5."title"::text) LIKE UPPER(%s)))
 ORDER BY "filtered_relation_book"."title" ASC,
-         T4."title" ASC
+         T5."title" ASC
```

This one is a bug, the filtered relation alias should be reuse.

```diff
--- test_with_m2m (filtered_relation.tests.FilteredRelationTests):postgresql:01ae9d4ca9
+++ test_with_m2m (filtered_relation.tests.FilteredRelationTests):postgresql:d7b1aba90a
@@ -5,5 +5,5 @@
 FROM "filtered_relation_author"
 INNER JOIN "filtered_relation_author_favorite_books" ON ("filtered_relation_author"."id" = "filtered_relation_author_favorite_books"."author_id")
 INNER JOIN "filtered_relation_book" favorite_books_written_by_jane ON ("filtered_relation_author_favorite_books"."book_id" = favorite_books_written_by_jane."id"
-                                                                       AND ("filtered_relation_author_favorite_books"."book_id" IN (%s)))
+                                                                       AND (favorite_books_written_by_jane."id" IN (%s)))
 WHERE favorite_books_written_by_jane."id" IS NOT NULL
```

This one is equivalent as `"filtered_relation_author_favorite_books"."book_id" = favorite_books_written_by_jane."id"` per the `JOIN` condition.

**Comment by charettes** [comment]
PR should be ready for review. The SQL diff for the `queries` and `filtered_relations` tests is

```diff
--- test_exclude_relation_with_join (filtered_relation.tests.FilteredRelationTests):postgresql:fb535e0a90
+++ test_exclude_relation_with_join (filtered_relation.tests.FilteredRelationTests):postgresql:ec481b0258
@@ -4,5 +4,6 @@
                 "filtered_relation_author"."object_id"
 FROM "filtered_relation_author"
 INNER JOIN "filtered_relation_book" book_alice ON ("filtered_relation_author"."id" = book_alice."author_id"
-                                                   AND (NOT (UPPER(book_alice."title"::text) LIKE UPPER(%s))))
+                                                   AND (NOT (UPPER(book_alice."title"::text) LIKE UPPER(%s)
+                                                             AND book_alice."title"::text IS NOT NULL)))
 WHERE book_alice."id" IS NOT NULL
```

The reason for the addition of this superflous `AND book_alice."title"::text IS NOT NULL` statement is that since filtered relation conditions are now resolved on first references instead of at the last minute the `filter(book_alice__isnull=False)` is not applied yet (it's the `book_alice` resferences that triggers the condition recolsing) so the join is still `LEFT OUTER` when it reaches the logic in charge of determining if an `IS NOT NULL` should be added.

The query remains the same so I don't think it's a huge deal if this gets added as it will only happen under rare circumstances (negated, demoted via `IS NOT NULL` filtering).

```
--- test_with_m2m (filtered_relation.tests.FilteredRelationTests):postgresql:fb535e0a90
+++ test_with_m2m (filtered_relation.tests.FilteredRelationTests):postgresql:ec481b0258
@@ -5,5 +5,5 @@
 FROM "filtered_relation_author"
 INNER JOIN "filtered_relation_author_favorite_books" ON ("filtered_relation_author"."id" = "filtered_relation_author_favorite_books"."author_id")
 INNER JOIN "filtered_relation_book" favorite_books_written_by_jane ON ("filtered_relation_author_favorite_books"."book_id" = favorite_books_written_by_jane."id"
-                                                                       AND ("filtered_relation_author_favorite_books"."book_id" IN (%s)))
+                                                                       AND (favorite_books_written_by_jane."id" IN (%s)))
 WHERE favorite_books_written_by_jane."id" IS NOT NULL
```

This one is equivalent as `"filtered_relation_author_favorite_books"."book_id" = favorite_books_written_by_jane."id"` per the JOIN condition.

**Comment by felixxm** [comment]
@charettes Thanks :heart:  ticket-33929 and ticket-34229 have been closed as duplicates of this issue, so we should add tests for them.

**Comment by charettes** [comment]
@felixxm I pushed some tests for both tickets but the MTI case is really complex to solve (hence the failing test)

Any objections to merging the first to commits and re-opening ticket-34229 to focus on the MTI case and using the third commit as a regression test?

**Comment by felixxm** [comment]
@charettes Looks great at first glance, thanks :star: I hope you don't work so late because of me :wink: :bed: 

> @felixxm I pushed some tests for both tickets but the MTI case is really complex to solve (hence the failing test)
> 
> Any objections to merging the first to commits and re-opening [ticket-34229](https://code.djangoproject.com/ticket/34229) to focus on the MTI case and using the third commit as a regression test?

No :+1: Sounds good.

**Comment by charettes** [comment]
@felixxm ticket-33929 is covered by this test. Aggregation has nothing to do with it, it was a variant of ticket-33766 it's just that the rhs was spawning JOIN against another `FilteredRelation`

https://github.com/django/django/blob/965917cd2616eec479722b575b6666c910363df3/tests/filtered_relation/tests.py#L811-L827

I pushed another commit that also removes the unnecessary `build_filtered_relation_q` method which was a copy-pasta of `_add_q` with disabled join type updates.

To summarize, first commit solves the issue for ticket-33766 and ticket-33929, second and third perform cleanups, fourth is a demonstration of the variant that warrants re-opening [ticket-34229](https://code.djangoproject.com/ticket/34229). LMK if that doesn't make sense.

File: django/db/models/query_utils.py [hunk 2]
```diff
@@ -418,18 +421,26 @@ def __eq__(self, other):
...
-        raise NotImplementedError("FilteredRelation.resolve_expression() is unused.")
+    def relabeled_clone(self, change_map):
+        clone = self.clone()
+        if resolved_condition := clone.resolved_condition:
+            clone.resolved_condition = resolved_condition.relabeled_clone(change_map)
+        return clone
+
+    def resolve_expression(self, query, reuse, *args, **kwargs):
+        clone = self.clone()
+        clone.resolved_condition = query.build_filter(
```
Comment (charettes) [inline django/db/models/query_utils.py:436]: We could use `_add_q` directly instead but `build_filter` is a public interface.

File: django/db/models/sql/query.py [hunk 3]
```diff
@@ -1042,6 +1047,18 @@ def join(self, join, reuse=None, reuse_with_filtered_relation=False):
...
         self.alias_map[alias] = join
+        if filtered_relation := join.filtered_relation:
+            resolve_reuse = reuse
+            if resolve_reuse is not None:
+                resolve_reuse = set(reuse) | {alias}
+            joins_len = len(self.alias_map)
+            join.filtered_relation = filtered_relation.resolve_expression(
+                self, reuse=resolve_reuse
+            )
+            # Some joins we're added by resolve expression, they must be present
```
Comment (charettes) [inline django/db/models/sql/query.py:1058]:
```suggestion
            # Some joins were added during expression resolve, they must be present
```

**Comment by felixxm** [comment]
@charettes Marvelous :clap: :star2: Thanks :1st_place_medal:

**Comment by felixxm** [comment]
I added test for `FilteredRelation` with self reference, see [comment](https://code.djangoproject.com/ticket/33766#comment:11).

**Comment by charettes** [comment]
> I added test for FilteredRelation with self reference

Thanks, I missed that one!
